### PR TITLE
[PAL/Linux-SGX] Add exitless system calls (new version)

### PR DESCRIPTION
**Pull request by dimakuv**
## Affected components

- [ ] README and global configuration
- [ ] Linux PAL
- [x] SGX PAL
- [ ] FreeBSD PAL
- [x] Common PAL code
- [ ] Library OS (i.e., SHIM), including GLIBC

## Description of the changes 

This PR succeeds PR #405. It is largely based on the old PR, but is rebased to the current master, and thus some fixes are not needed anymore. No logic was changed.

This PR adds the ability to invoke system calls in an "exitless" fashion (aka asynchronously), similar to solutions from [Eleos by Orenbach](https://dl.acm.org/citation.cfm?id=3064219) and [SCONE by Arnautov et al](https://dl.acm.org/citation.cfm?id=3026930).

New directive `sgx.rpc_thread_num = X` in the manifest instructs Graphene-SGX to create X outside-enclave RPC threads that spin-wait for syscalls (and any OCALLs in general) on a shared queue. In-enclave threads do not exit the enclave now but instead enqueue the syscall in the queue and wait for it to be processed by an RPC thread. Not to waste CPU resources, enclave threads first spin (in the hope syscall will return immediately) and then sleep on a futex.

This PR reserves the SIGUSR2 signal for internal usage. In particular, when a signal arrives to the enclave application (e.g. SIGINT or SIGCONT), RPC threads are notified using SIGUSR2. This notification is required in case RPC threads are blocked on system calls; POSIX mandates that signals interrupt such blocked system calls ([`signal` manpage](http://man7.org/linux/man-pages/man7/signal.7.html)).

Omitting `sgx.rpc_thread_num` defaults to the old way of executing syscalls: exiting the enclave, executing the syscall, and re-entering it again.

This PR is stable. It works with multi-process applications and correctly handles signals.

Closes #405.

## How to test this PR? 

Two new regression tests are added (one in LibOS, one in PAL).

---
This change is [Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1142)

### CODE REVIEW

**Comment by dimakuv** [comment]
## Performance numbers

Quick performance evaluation. I used an SGX v1 machine with Intel(R) Xeon(R) CPU E3-1280 v5 @ 3.70GHz, 8 logical CPUs, 8MB LLC cache, and 48GB of RAM, with Ubuntu 18.04 and Linux 5.1.

**TLDR**: Exitless improves performance **~2X**.

### Redis

Redis is (almost) single-threaded. Our example manifest creates 8 enclave threads, so I added 8 RPC threads. Most of them sleep anyway and do not affect performance.

Running our example like this: `$ SGX=1 ./pal_loader redis-server --save '' --protected-mode no`.

- without Exitless:
```
$ src/src/redis-benchmark -t get,set -q
SET: 11018.07 requests per second
GET: 11181.93 requests per second
```

- with Exitless:
```
$ src/src/redis-benchmark -t get,set -q
SET: 24636.61 requests per second
GET: 25000.00 requests per second
```

### Memcached

Memcached is weirdly multi-threaded. Even when running with 2 threads, it creates another ~8 threads for some bookkeeping. Thus, I added 10 RPC threads. The resulting configuration occupies all my 8 logical CPUs, each with almost 100% utilization.

Running our example like this: `$ SGX=1 ./pal_loader memcached -u nobody -t 2`. I restricted to two threads since I only have a 8-logical CPUs machine.

- without Exitless:
```
$ memtier_benchmark --port=11211 --protocol=memcache_binary --hide-histogram
Type         Ops/sec     Hits/sec   Misses/sec      Latency       KB/sec
-------------------------------------------------------------------------
Sets         2570.12          ---          ---     45.21000       197.94
Gets        25673.00         0.00     25673.00      0.17000       975.00
Totals      28243.13         0.00     25673.00      4.26900      1172.95
```

- with Exitless:
```
$ memtier_benchmark --port=11211 --protocol=memcache_binary --hide-histogram
Type         Ops/sec     Hits/sec   Misses/sec      Latency       KB/sec
-------------------------------------------------------------------------
Sets         4946.72          ---          ---     16.27800       380.98
Gets        49412.89         0.00     49412.89      1.52600      1876.59
Totals      54359.61         0.00     49412.89      2.86800      2257.57
```

**Comment by dimakuv** [comment]
Jenkins, retest Jenkins-Debug please

**Comment by dimakuv** [comment]
Jenkins, retest Jenkins-18.04 please (`futex_wait_bitset01` LTP test failed)
